packages/flutter/lib/src/painting/debug_overflow_indicator.dart

```diff
@@ -0,0 +1,392 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+import 'dart:ui' as ui;
+
+import 'package:flutter/rendering.dart';
+
+// Data used by the DebugOverflowIndicator to manage the regions and labels for
+// the indicators.
+class _OverflowRegionData {
+  _OverflowRegionData({
+    this.rect,
+    this.label = '',
+    this.labelOffset = Offset.zero,
+    this.rotation = 0.0,
+  });
+
+  final Rect rect;
+  final String label;
+  final Offset labelOffset;
+  final double rotation;
+}
+
+/// An mixin indicator that is drawn when a [RenderObject] overflows its
+/// container.
+///
+/// This is used by RenderObjects that are containers to show where, and by how
+/// much, their children overflow their containers.  These indicators are
+/// typically only shown in a debug build (where the call to
+/// [overflowPaintIndicator] is surrounded by an assert).
+///
+/// Typical usage is to set the [overflowContainerRect] and the
+/// [overflowChildRect], and then call [overflowPaintIndicator]. This will
+/// calculate the overflow, and then draw the relevant indicators.
+///
+/// This class will also print a debug message to the console when the container
+/// overflows, but only on the first occurrence.  To reset and print it again
+/// upon the next occurrence, set [overflowReportNeeded] to true. This is
+/// typically done in the [RenderObject.reassemble] override in the
+/// [RenderObject] that uses the DebugOverflowIndicatorMixin.
+///
+/// ## Sample code
+///
+/// ```dart
+/// class MyRenderObject extends RenderObject with DebugOverflowContainerMixin {
+///   @override
+///   TextDirection get textDirection => _textDirection;
+///
+///   @override
+///   void performLayout() {
+///     // ...
+///     overflowChildRect = childParentData.offset & child.size;
+///     overflowContainerRect = Offset.zero & size;
+///   }
+///
+///   @override
+///   void paint(PaintingContext context, Offset offset) {
+///     if (isOverflowing) {
+///       _paintStuff(context, offset);
+///       return;
+///     }
+///     // ...
+///     assert(() {
+///       overflowRenderObject ??= this;
+///       overflowPaintIndicator(context, offset);
+///       return true;
+///     });
+///   }
+///
+///   @override
+///   void reassemble() {
+///     super.reassemble();
+///     assert(() {
+///       overflowReportNeeded = true;
+///       return true;
+///     }());
+///   }
+/// }
+/// ```
+///
+/// See also:
+///   * The code for [RenderUnconstrainedBox], and [RenderFlex] for examples of
+///     classes that use this indicator mixin.
+class DebugOverflowIndicatorMixin {
```

Comment:
This class is named DebugFoo, but its methods are not limited to debug mode.
I think each method should be wrapped in an assert so that in non-debug builds, all these calls are no-ops.